floodpipeline/forecast.py

```diff
@@ -0,0 +1,601 @@
+                        gdf_adm,
+                        raster_array,
+                        affine=transform,
+                        stats=["sum"],
+                        all_touched=True,
+                        nodata=0.0,
+                    )
+                    gdf_pop = pd.concat([gdf_adm, pd.DataFrame(stats)], axis=1)
+                    gdf_pop.index = gdf_pop[f"adm{adm_lvl}_pcode"]
+
+                # add affected population to forecast data units
+                for forecast_data_unit in self.data.forecast_admin.get_data_units(
+                    adm_level=adm_lvl, lead_time=lead_time
+                ):
+                    if forecast_data_unit.triggered:
+                        try:
+                            pop_affected = int(
+                                gdf_aff_pop.loc[forecast_data_unit.pcode, "sum"]
+                            )
+                        except (ValueError, TypeError, KeyError):
+                            pop_affected = 0
+                        forecast_data_unit.pop_affected = pop_affected
+                        try:
+                            forecast_data_unit.pop_affected_perc = (
+                                float(
+                                    pop_affected
+                                    / gdf_pop.loc[forecast_data_unit.pcode, "sum"]
+                                )
+                                * 100.0
+                            )
+
+                        except (ValueError, TypeError, KeyError):
+                            forecast_data_unit.pop_affected_perc = 0.0
+
+    def compute_forecast_station(self):
+        """
+        Forecast floods per GloFAS station based on river discharge data
+        1. determine if trigger level is reached, with which probability, and alert class
+        """
+        self.__compute_triggers_station()
+
+    def __compute_triggers_station(self):
+        """Determine if trigger level is reached, its probability, and the alert class"""
+
+        os.makedirs(self.input_data_path, exist_ok=True)
+        os.makedirs(self.output_data_path, exist_ok=True)
+        country = self.data.forecast_station.country
+
+        trigger_on_lead_time = self.settings.get_country_setting(
+            country, "trigger-on-lead-time"
+        )
+        for discharge_station in self.data.discharge_station.data_units:
+            station_code = discharge_station.station_code
+            lead_time = discharge_station.lead_time
+            threshold_station = self.data.threshold_station.get_data_unit(station_code)
+
+            likelihood_per_return_period, forecasts = {}, []
+            
+            for threshold in threshold_station.thresholds:
+                threshold_checks = map(
+                    lambda x: 1 if x > threshold["threshold_value"] else 0,
+                    discharge_station.discharge_ensemble,
+                )
+                likelihood = sum(threshold_checks) / len(
```

Comment:
This introduces the risk of a divide-by-zero error. [Here are a few ways to resolve the error](https://stackoverflow.com/a/29836987/1753041).